README.md
Spam Rule is a plugin for SquirrelMail that lets you select from different
groups of email filters via a config file maintained through FTP.  It is
written specifically for situations where a user can select from various
filters, individually or in predefined groups, and subscribe to them by
modifing lines in a procmail or maildrop config file in their home
directory.  This way the user can have his mail filtered at message delivery
time by the email system rather than at reading time by SquirrelMail, which
could really slow things down.

See also:  http://www.squirrelmail.org/plugin_view.php?id=190


## Requirements

- access to an FTP server for each IMAP user, using the same login name
- a forward file somewhere under the user's FTP directory,
  e.g. ".forward" or ".qmail"
- a mail filter config file somewhere under the user's FTP directory,
  e.g. ".mailfilter" for maildrop or ".procmailrc" for procmail
- version 2.x of the compatibility plugin


## Features

- creates a new block under Options called "Spam Rules"
- tries to give the user a simple interface that can still provide as much
  flexibility as the user desires
- lets the admin create sets of predefined filter groups from which the
  user can select, but also lets the user pick and choose from the whole
  set of available filters
- tries to be very flexible (perhaps too much so?)
- support for translations


## Installation

- Install this directory into your squirrelmail source dir at
  plugins/spamrule.
- Copy config.php.ex to config.php and modify the file as described within.
  You may also want to leave this file intact and simply put your own
  variable settings into config_local.php instead, which will override
  anything defined in config.php.
- Take a look at the maildrop subdirectory.  There's a sample of my
  system-wide maildrop config, /usr/local/etc/maildroprc, and individual
  rule files that get installed under /usr/local/etc/maildroprcs.  This
  plugin will include the ones that a user selects by customizing his
  .mailfilter file.  You can add your own by changing config.php, but
  it's already set up to take advantage of the ones I've included.
- Enable the plugin through conf.pl.
- Test it out!  Make sure to double-check the contents of your forward file
  and mail filter config file, as well as successful results of message
  filtering and delivery, before putting it into production.


## Upgrading

- If you've used spamrule in the past, you can now override config.php's
  settings in config_local.php.  Any variables defined in config.php
  can be put there.  Since it's loaded after config.php, your settings
  will override the defaults.
- Release 0.4 added internationalization and included major updates to
  the strings in config.php.  I strongly recommend that you don't just
  keep your old pre-0.4 config, but rather copy config.php.ex to
  config.php and add your new settings there or to config_local.php.


## Questions and support

I've tested this plugin quite a bit and it has worked fantastic on my
systems since 2002.  I have tried to make it resistant to unexpected
problems and easy to change for different environments, but I'm sure
there are many shortcomings.  If you have any questions or comments,
please contact me through the project's Github page at
https://github.com/oshaughnessy/squirrelmail-spamrule.


## Copyright

Copyright (c) 2002-2013 O'Shaughnessy Evans <shaug-sqml @ wumpus.org>


## License

This code is licensed under the Perl Artistic License, version 2.0.  For
more information, please see the file Artistic_2.0, which was included with
this distribution, or http://opensource.org/licenses/artistic-license-2.0.php


NEWS
Version 0.5 / Sept 13, 2007
  - made the translation support much more sensible
  - resolved all the PHP warnings
  - added compatibility plugin back in, using compatibility 2.x now.
  - added example files:  config.php.ex and config_local.php.ex, and 
    the ability to override config.php in config_local.php

Version 0.4 / Sept 11, 2007
  - updated code to follow recommended standards for squirrelmail plugins
    based on guidelines at http://squirrelmail.org/docs/devel/devel-2.html
  - added support for translations
  - fixed a bug in the list editor where it couldn't add new entries to
    a file if the file didn't already exist

Version 0.3 / Oct 3, 2003
  - added support for squirrelmail 1.4
  - added a "Spam Rules" entry to the menuline (top of right frame)
  - switched from reading $GLOBALS, $_GET, etc., to using sqgetGlobalVar
  - cleaned up the UI a little
  - fixed a bug where backslashes could start to grow exponentially in the
    maildrop list files
  - switched config.php defaults from supporting qmail to sendmail/postfix
    (syntax of the .forward file is slightly different than .qmail)

Version 0.2
  - first public release


options.php
<?php

/**
 * spamrule/options.php
 * 
 * Options page for the SquirrelMail "spamrule" plugin,
 * which provides users with simple management of email filter groups.
 * 
 * @copyright Copyright (c) 2002-2007 O'Shaughnessy Evans <shaug-sqml @ wumpus.org>
 * @version $Id: options.php,v 1.9 2007/09/13 19:59:21 shaug Exp $
 * @license http://opensource.org/licenses/artistic-license-2.0.php
 * @package plugins
 * @subpackage spamrule
 */


// load init scripts for SquirrelMail 1.5 or 1.4
if (file_exists('../../include/init.php')) {
    require('../../include/init.php');
}
else if (file_exists('../../include/validate.php')) {
    if (!defined('SM_PATH')) {
        define('SM_PATH', '../../');
    }
    include_once(SM_PATH . 'include/validate.php');
} 
else if (file_exists('../../src/validate.php')) {
    chdir('..');
    if (!defined('SM_PATH')) {
        define('SM_PATH', '../');
    }
    include_once(SM_PATH . 'src/validate.php');
}

// load the compatibility plugin
include_once(SM_PATH . 'plugins/compatibility/functions.php');

global $SPAMRULE_OPTS, $action, $color;
// When letting the user select from the CUSTOM group of filters,
// we're merging data from the form into SPAMRULE_OPTS[groups][CUSTOM],
// so we can't use sqgetGlobalVar to manipulate that variable here.
//sqgetGlobalVar('SPAMRULE_OPTS', $SPAMRULE_OPTS);
sqgetGlobalVar('action', $action, SQ_FORM);


if ($action != 'editlist') {
    displayPageHeader($color, '');
}


// we're internationalized, so bind gettext functions to our domain
textdomain('spamrule');


// load plugin configs and functions
load_config('spamrule', array('config.php'));
if (file_exists(SM_PATH . 'plugins/spamrule/config_local.php')) {
    include_once(SM_PATH . 'plugins/spamrule/config_local.php');
}
require_once(SM_PATH . 'plugins/spamrule/lib.php');


echo <<<EOjs

<script language="JavaScript" type="text/javascript">
  // subwin:  a simple function to open a new window
  // arguments:
  //   uri      page that the new window should load
  //   name     window target name
  //   w        width of the window
  //   h        height of the window
  function subwin(uri, name, w, h) {
      window.open(uri, name, "width="+w+",height="+h+",resizable,"+
                             "scrollbars,status");
  }
</script>


EOjs;


switch ($action) {
case 'groups':
    sqgetGlobalVar('newgroup', $group, SQ_FORM);
    sqgetGlobalVar('finish', $finish, SQ_FORM);

    if ($group == 'CUSTOM' && !isset($finish)) {
     // !isset($_POST['SPAMRULE_OPTS']['groups'][$group]['filters'])) {
        sr_print_header($SPAMRULE_OPTS['custom']['header'], 3);
        sr_select_filters();
        sr_print_footer($SPAMRULE_OPTS['custom']['footer'], 3);
    }
    else {
        $filters = $group == 'CUSTOM'
                   ? $_POST['SPAMRULE_OPTS']['groups'][$group]['filters']
                   : $SPAMRULE_OPTS['groups'][$group]['filters'];
        sr_install_filters($group, $filters); // || fail();
        //sr_print_header($SPAMRULE_OPTS['groups']['summary']['header'], 3);
        //sr_print_summary($group);
        //sr_print_footer($SPAMRULE_OPTS['groups']['summary']['footer'], 3);
        sr_print_header($SPAMRULE_OPTS['show']['header'], 3);
        sr_print_subscribed();
        sr_print_footer($SPAMRULE_OPTS['show']['footer'], 3);
    }
    break;
case 'custom':
    break;
case 'subscribed':
    sr_print_header($SPAMRULE_OPTS['show']['header'], 3);
    sr_print_subscribed();
    sr_print_footer($SPAMRULE_OPTS['show']['footer'], 3);
    break;
case 'editlist':
    sqgetGlobalVar('list', $list, SQ_FORM);
    displayHtmlHeader(_("Spam Rules:"). ' '. $SPAMRULE_OPTS[$list]['title']);
    switch ($list) {
    case 'allowed_senders':
    case 'blocked_senders':
    case 'blocked_subjects':
    case 'subject_passwords':
    case 'allowed_recipients':
        sr_print_header($SPAMRULE_OPTS[$list]['header'], 3);
        sr_maintain_list($list);
        sr_print_footer($SPAMRULE_OPTS[$list]['footer'], 3);
        break;
    default:
        sr_print_header($SPAMRULE_OPTS['groupsel']['header'], 3);
        sr_select_group();
        sr_print_footer($SPAMRULE_OPTS['groupsel']['footer'], 3);
        break;
    }
    break;
default:
    sr_print_header($SPAMRULE_OPTS['groupsel']['header'], 3);
    sr_select_group();
    sr_print_footer($SPAMRULE_OPTS['groupsel']['footer'], 3);
    break;
}

?>

</body>
</html>


config_local.php.ex
<?php

/*
 * spamrule/config_local.php
 * 
 * If you want to override any settings in config.php, put them here.
 * This file will not be overwritten during upgrades.
 */

//$SPAMRULE_OPTS['ftphost'] = "ftp.$domain";

//$SPAMRULE_OPTS['filters']['GLOBAL'] = NULL;
//$SPAMRULE_OPTS['filters']['spamassassin'] = NULL;

//$SPAMRULE_OPTS['filters']['GLOBAL']['text'] .= '  '.
//    _("If you would like to learn more about these filters, ".
//      "please see"). ' '.
//    "<a href=\"http://www.$domain/support/antispam.php\" target=\"_blank\">".
//    _("our Anti-Spam page"). '</a>.';

//$SPAMRULE_OPTS['filters']['GLOBAL']['install'] =
// '/usr/local/www/libexec/spamfriend-del '. $username;

//$SPAMRULE_OPTS['filters']['GLOBAL']['uninstall'] =
// '/usr/local/www/libexec/spamfriend-add '. $username;

?>


setup.php
<?php

/**
 * spamrule/setup.php
 * 
 * The plugin initialization page for the SquirrelMail "spamrule" plugin,
 * which provides users with simple management of email filter groups.
 * 
 * @copyright Copyright (c) 2002-2007 O'Shaughnessy Evans <shaug-sqml @ wumpus.org>
 * @version $Id: setup.php,v 1.8 2007/09/13 18:56:07 shaug Exp $
 * @license http://opensource.org/licenses/artistic-license-2.0.php
 * @package plugins
 * @subpackage spamrule
 */


/**
 * squirrelmail_plugin_init_spamrule()
 * 
 * Initialize the plugin.
 */
function squirrelmail_plugin_init_spamrule()
{
    global $squirrelmail_plugin_hooks;

    $squirrelmail_plugin_hooks['optpage_register_block']['spamrule']
     = 'spamrule_options';
    $squirrelmail_plugin_hooks['menuline']['spamrule'] = 'spamrule_menuline';
}


/**
 * spamrule_options()
 * 
 * Set up the Options page block.
 */
function spamrule_options()
{
    global $optpage_blocks;

    //bindtextdomain('spamrule', SM_PATH. 'plugins/spamrule/locale');
    textdomain('spamrule');

    $optpage_blocks[] = array(
        'name' => _("Spam Rules"),
        'url'  => '../plugins/spamrule/options.php',
        'desc' => _("Decide how you want to control the incoming flood of ".
                    "mail you didn't ask for.  These filters work all the ".
                    "time, whether you use webmail or your own program to ".
                    "read email."),
        'js'   => FALSE
    );

    //bindtextdomain('squirrelmail', SM_PATH. 'locale');
    textdomain('squirrelmail');
}


/**
 * spamrule_menuline()
 * 
 * Add a link to the main frame's menu line.
 */
function spamrule_menuline()
{
    //bindtextdomain('spamrule', SM_PATH. 'plugins/spamrule/locale');
    textdomain('spamrule');

    displayInternalLink('plugins/spamrule/options.php', _("Spam Filters"), '');
    echo '&nbsp;&nbsp;';

    //bindtextdomain('squirrelmail', SM_PATH. 'locale');
    textdomain('squirrelmail');
}


/**
 * spamrule_info()
 * 
 * @returns array with various bits of information about the plugin.
 * Thanks to Paul Lesniewski for the example in the compatibility plugin.
 * 
 * Each element in the array is an info parameter.  Elements may be of any type.
 */
function spamrule_info()
{
    return array(
        'english_name' => 'Spamrule',
        'authors' => array(
            'O\'Shaughnessy Evans' => array('email' => 'shaug-sqml@wumpus.org'),
        ),
        'version' => '0.5',
        'required_sm_version' => '1.2',
        'requires_configuration' => 1,
        'requires_source_patch' => 0,
        'required_plugins' => array('compatibility'),
        'summary' => 'Provides users with simple sets of filters from which '.
                     'they can choose, or lets them customize their own set.',
        'details' => <<<EOdetails
Spamrule gives users an Options page to configure their own set of spam filters based on maildrop or procmail filters provided by the site admin. The filters can be grouped however you like (e.g. "no filters", "loose filters", and "strict filters") so that users can just choose a group, or users can choose to select from any of the filters to create their own group if they like. There are also mechanisms to edit lists that the filters use -- e.g. a filter might only allow email from certain senders, then the list editor page would let them maintain that list of senders. It's designed to initially appear as simple as Hotmail's filters, but it provides much more flexibility if the user wants it.
EOdetails
    );
}


/**
 * spamrule_version()
 * 
 * @returns string identifying the plugin's version number.
 */
function spamrule_version()
{
    $info = spamrule_info();
    return $info['version'];
}
